File: rest-service/src/main/java/mb/api/controller/resources/Reports.java
package mb.api.controller.resources;

import java.util.List;
import javax.ws.rs.*;
import javax.ws.rs.core.Response;

import mb.api.model.reports.ReportRequest;
import mb.api.model.reports.ReportResponse;
import mb.lib.http.MimeType;

@Path(Paths.Reports)
public interface Reports
{
  /**
   * GET /reports
   */
  @GET
  @Produces(MimeType.ApplicationJSON)
  List<ReportResponse> getAllReports(@QueryParam(Vars.JobID) String jobID);

  /**
   * POST /reports
   */
  @POST
  @Produces(MimeType.ApplicationJSON)
  ReportResponse newReport(ReportRequest req);

  /**
   * GET /reports/{report-id}
   */
  @GET
  @Path(Paths.ReportByID)
  @Produces(MimeType.ApplicationJSON)
  ReportResponse getReport(@PathParam(Vars.ReportID) String reportID);

  /**
   * POST /reports/{report-id}
   */
  @POST
  @Path(Paths.ReportByID)
  @Produces(MimeType.ApplicationJSON)
  ReportResponse rerunReport(@PathParam(Vars.ReportID) String reportID);

  /**
   * GET /reports/{report-id}/files/{filename}
   */
  @GET
  @Path(Paths.ReportData)
  @Produces({
    MimeType.ApplicationBinary,
    MimeType.ApplicationJSON,
    MimeType.ApplicationXML,
    MimeType.ApplicationZip,
    MimeType.TextPlain
  })
  Response getReportData(
    @PathParam(Vars.ReportID) String reportID,
    @PathParam(Vars.FileName) String fileName,
    @QueryParam("download") @DefaultValue("true") boolean download
  );
}

File: rest-service/src/main/kotlin/mb/api/controller/resources/Reports.kt
package mb.api.controller.resources

import mb.api.model.io.Headers
import mb.api.model.reports.ReportRequest
import mb.api.model.reports.ReportResponse
import mb.lib.http.MimeType
import javax.ws.rs.*
import javax.ws.rs.core.Response

@Path(Paths.Reports)
interface Reports
{
  /**
   * GET /reports
   */
  @GET
  @Produces(MimeType.ApplicationJSON)
  fun getAllReports(@QueryParam(Vars.JobID) jobID: String?): List<ReportResponse>

  /**
   * POST /reports
   */
  @POST
  @Produces(MimeType.ApplicationJSON)
  fun newReport(req: ReportRequest): ReportResponse

  /**
   * GET /reports/{report-id}
   */
  @GET
  @Path(Paths.ReportByID)
  @Produces(MimeType.ApplicationJSON)
  fun getReport(@PathParam(Vars.ReportID) reportID: String): ReportResponse

  /**
   * POST /reports/{report-id}
   */
  @POST
  @Path(Paths.ReportByID)
  @Produces(MimeType.ApplicationJSON)
  fun rerunReport(@PathParam(Vars.ReportID) reportID: String): ReportResponse

  /**
   * GET /reports/{report-id}/files/{filename}
   */
  @GET
  @Path(Paths.ReportData)
  @Produces(value = [
    MimeType.ApplicationBinary,
    MimeType.ApplicationJSON,
    MimeType.ApplicationXML,
    MimeType.ApplicationZip,
    MimeType.TextPlain
  ])
  fun getReportData(
    @PathParam(Vars.ReportID) reportID: String,
    @PathParam(Vars.FileName) fileName: String,
    @QueryParam("download") @DefaultValue("true") download: Boolean,
    @HeaderParam(Headers.ContentMaxLength) contentMaxLength: Long?,
  ): Response
}
